In [1]:
pip install google-api-python-client

You should consider upgrading via the '/Users/lidang/opt/anaconda3/bin/python -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [1]:
from googleapiclient.discovery import build
my_api_key = 'AIzaSyB5tKEY1y1v7VdQTeDQWlZPlPBI9UPp_Lc'
my_cse_id = '154b6a50a71a04d3a'

def google_search(search_term, api_key, cse_id, **kwargs):
    service = build("customsearch", "v1", developerKey=api_key)
    res = service.cse().list(q=search_term, cx=cse_id, **kwargs).execute()
    return res

In [2]:
result = google_search('site:linkedin.com/in/ AND "Richard Aaronson" AND "University of Chicago"', my_api_key, my_cse_id)
print(result['items'])

[{'kind': 'customsearch#result', 'title': 'Rich Aaronson - Consultant / Social Entrepreneur - Various projects ...', 'htmlTitle': 'Rich Aaronson - Consultant / Social Entrepreneur - Various projects ...', 'link': 'https://www.linkedin.com/in/rich-aaronson-973a9028', 'displayLink': 'www.linkedin.com', 'snippet': 'University of Chicago. MASocial Sciences. University of Pennsylvania University of ... Richard Aaronson. CEO at Atlantic Realty Partners. Atlanta, GA\xa0...', 'htmlSnippet': '<b>University of Chicago</b>. MASocial Sciences. University of Pennsylvania University of ... <b>Richard Aaronson</b>. CEO at Atlantic Realty Partners. Atlanta, GA&nbsp;...', 'formattedUrl': 'https://www.linkedin.com/in/rich-aaronson-973a9028', 'htmlFormattedUrl': 'https://www.linkedin.com/in/rich-aaronson-973a9028', 'pagemap': {'cse_thumbnail': [{'src': 'https://encrypted-tbn3.gstatic.com/images?q=tbn:ANd9GcRsk8Bdn-BiMCZD6JvGq4P1btm-3wsl7PWrC6nSUcumUxBUqxWc5D1fRWSJ', 'width': '225', 'height': '225'}], 'me

In [56]:
import pandas as pd

df = pd.read_excel('MA graduate data.xlsx')
df = df.rename(columns={"Person::NAME_First": "f_name", "Person::NAME_Last": "l_name"})
df

ID_UCID     f_name     l_name Person::GENDER_Code Person::BIRTH_Date  \
0     10916201  Katherine     Bukolt                   F         1971-05-27   
1     10914988      Isaac  Greenspan                   M         1979-11-20   
2     10891647       Carl  Malmstrom                   M         1977-12-28   
3     10909591    Richard   Aaronson                   M         1960-12-27   
4     10916706    Haytham   Abu Zayd                   M         1978-08-22   
...        ...        ...        ...                 ...                ...   
5125  12333428    Melissa      Yeung                   F         1999-08-14   
5126  12331953    Wenqing         Yu                   F         1999-01-23   
5127  12332609      Xuran       Zeng                   F         1999-01-31   
5128  12278012      Bohan      Zhang                   M         1998-01-21   
5129  12332687      Zihan        Zhu                   M         1999-01-20   

      Person::CTZ_status  ACAD_cohort                   DEPT_Code.AIS  \
0           U.S. Citizen         2000   MESS - Middle Eastern Studies   
1           U.S. Citizen         2000                EDUM - Education   
2           U.S. Citizen         2000                    MAPS - MAPSS   
3           U.S. Citizen         2000                    MAPS - MAPSS   
4           U.S. Citizen         2000                    MAPS - MAPSS   
...                  ...          ...                             ...   
5125  non-resident alien         2021                    MAPS - MAPSS   
5126  non-resident alien         2021                    MAPS - MAPSS   
5127  non-resident alien         2021                    MAPS - MAPSS   
5128  non-resident alien         2021  INRE - International Relations   
5129  non-resident alien         2021                    MAPS - MAPSS   

                          Person::COLL_InstName Person::COLL_Degr MA_DegrQtr  \
0                      University of Notre Dame              B.A.    2003/01   
1                           New York University              B.A.    2001/02   
2                         University of Chicago              B.A.    2001/02   
3                    University of Pennsylvania              B.S.    2002/01   
4                             McGill University              B.A.    2001/04   
...                                         ...               ...        ...   
5125                  University College London               NaN    2022/02   
5126  Chinese University Of Hong Kong, Shenzhen               BBA    2022/02   
5127                 Renmin University of China             BECON    2022/02   
5128                       University Of Oxford             M.St.    2022/02   
5129                           Colorado College                BA    2022/02   

     MA_degrDate  
0     2003-03-22  
1     2001-06-09  
2     2001-06-09  
3     2002-03-16  
4     2001-12-08  
...          ...  
5125  2022-06-04  
5126  2022-06-04  
5127  2022-06-04  
5128  2022-06-04  
5129  2022-06-04  

[5130 rows x 12 columns]

## To ensure we have enough quota, I have created two sets of api id and keys

my_api_key = 'AIzaSyD6u3kV9ZkgdmXa1SJL_esV2Nha_OEhWTM'

my_cse_id = '154b6a50a71a04d3a'



my_api_key = 'AIzaSyCjnheo33d36jveW_zF-Ntd-eT0AFZlC80'

my_cse_id = 'c6ad05d49d36a485d'

In [8]:
import json
import time
with open('api.json') as datafile:
    data = json.load(datafile)

In [120]:
from googleapiclient.discovery import build

my_api_key = data['my_api_key']
my_cse_id = data['my_cse_id']

def google_search(search_term, api_key, cse_id, **kwargs):
    service = build("customsearch", "v1", developerKey=api_key)
    res = service.cse().list(q=search_term, cx=cse_id, **kwargs).execute()
    return res
def search_result(name, my_api_key, my_cse_id, results):
    result = google_search(f'site:linkedin.com/in/ AND {name} AND "University of Chicago"', my_api_key, my_cse_id)
    num_result = int(result['searchInformation']['totalResults'])
    if num_result < 1:
        results[name_space] = ['no result']
        return 0
    else:
        title, url, snippet = result['items'][0]['title'], result['items'][0]['formattedUrl'], result['items'][0]['snippet']
        results[name_space] = [title, url, snippet]
        print(title)
        return 1
def search_multiple_result(name, my_api_key, my_cse_id, results):
    result = google_search(f'site:linkedin.com/in/ AND {name} AND "University of Chicago"', my_api_key, my_cse_id)
    num_result = int(result['searchInformation']['totalResults'])
    if num_result < 1:
        results[name_space] = ['no result']
        return 0
    else:
        results[name_space] = []
        for index, item in enumerate(result['items']):
            title, url, snippet = item['title'], item['formattedUrl'],item['snippet']
            results[name_space].append({index: [title, url, snippet]})
        return 1

In [133]:
def gather_results(sample):
    counter = 0
    for row in sample.itertuples():
        counter = counter + 1
        if counter % 100 == 0:
            print(counter)
            time.sleep(65)
        name_space = row.f_name + " " + row.l_name
        name = row.f_name + row.l_name
        results[name_space] = []
        title = ""
        url = ""
        snippet = ''
        result = google_search(f'site:linkedin.com/in/ AND {name} AND "University of Chicago"', my_api_key, my_cse_id)
        num_result = int(result['searchInformation']['totalResults'])
        if num_result < 1:
            results[name_space] = ['no result']
        else:
            title, url, snippet = result['items'][0]['title'], result['items'][0]['formattedUrl'], result['items'][0]['snippet']
            results[name_space] = [title, url, snippet]

After adding billing information, we could successfully siwtch to the custom search json API with daily quota of 10,000. 
Make sure to sleep for slightly longer than a minute to circumvent the minute quota everytime we finished 100 API calls

around 1200 results, there could easily be an internal error

Split the dataframe into 6 dataframes for data collection -- large dataframe (>1000) tend to fault at execution

In [168]:
df_1 = df.iloc[:1000,:]
df_2 = df.iloc[1000:2000]
df_3 = df.iloc[2000:3000]
df_4 = df.iloc[3000:4000]
df_5 = df.iloc[4000:5000]
df_6 = df.iloc[5000:,:]

## Get results for all alumni

In [136]:
results = {}

In [138]:
gather_results(df_1)

100
200
300
400
500
600
700
800
900
1000


In [139]:
gather_results(df_2)

100
200
300
400
500
600
700
800
900
1000


In [140]:
gather_results(df_3)

100
200
300
400
500
600
700
800
900
1000


In [170]:
gather_results(df_4)

100
200
300
400
500
600
700
800
900
1000


In [148]:
gather_results(df_5)

100
200
300
400
500
600
700
800
900
1000


In [149]:
gather_results(df_6)

100


With sleep(), 5135 queries take roughly took 1 hour and 30 minutes. From time to time the api malfunctions, so need to pay close attention to the tracker

In [172]:
len(results)

5120

There are duplicates in the data

In [165]:
sum(df.duplicated(subset=['f_name', 'l_name']))

10

So 5120 rows should be right

## Save to csv

In [228]:
import csv
 
with open('results_collection.csv', 'w') as csv_file:  
    writer = csv.writer(csv_file)
    for key, value in results.items():
        writer.writerow([key, value])

In [29]:
count_no_result = 0
for key, values in results.items():
    if values[0] != "no result":
        count_no_result = count_no_result + 1
print("no results: " + str(count) + "\n" + "have results: " + str((len(results) - count_no_result)))

no results: 2986
have results: 2134
